app/src/main/java/org/thoughtcrime/securesms/ApplicationContext.java

```diff
@@ -492,4 +494,45 @@ public void restartApplication() {
     }
 
     // endregion
+
+    public void wakeUpDeviceAndDismissKeyguardIfRequired() {
+        // Get the KeyguardManager and PowerManager
+        KeyguardManager keyguardManager = (KeyguardManager)getSystemService(Context.KEYGUARD_SERVICE);
+        PowerManager powerManager       = (PowerManager)getSystemService(Context.POWER_SERVICE);
+
+        // Check if the phone is locked & if the screen is awake
+        boolean isPhoneLocked = keyguardManager.isKeyguardLocked();
+        boolean isScreenAwake = powerManager.isInteractive();
+
+        if (!isScreenAwake) {
+            PowerManager.WakeLock wakeLock = powerManager.newWakeLock(
+                    PowerManager.FULL_WAKE_LOCK
+                            | PowerManager.ACQUIRE_CAUSES_WAKEUP
+                            | PowerManager.ON_AFTER_RELEASE,
+                    WAKELOCK_TAG);
+
+            // Acquire the wake lock to wake up the device
+            wakeLock.acquire(3000);
+
+            // Wait for the device to actually wake up
+            AtomicBoolean stopWaitingForWakeUp = new AtomicBoolean(false);
+            long MAX_WAIT_PERIOD_MS = 50L;
+            Handler handler = new Handler(Looper.getMainLooper());
+            handler.postDelayed(() -> stopWaitingForWakeUp.set(true), MAX_WAIT_PERIOD_MS);
+
+            // Busy-wait until the screen is interactive - to be safe, we'll wait a maximum of 50ms.
+            do {
+                 /* Typically this loop runs very briefly - for somewhere in the order of 10ms or less */
+            } while (!powerManager.isInteractive() && !stopWaitingForWakeUp.get());
+        }
```

Comment:
Unfortunately, adjusting the delay mechanism to:
```
// Wait for the device to actually wake up
long MAX_WAIT_PERIOD_MS = 50L;
final long start = SystemClock.uptimeMillis();
while (!powerManager.isInteractive() && SystemClock.uptimeMillis() - start <= MAX_WAIT_PERIOD_MS) {
    try {
        Thread.sleep(10);
    } catch (InterruptedException e) {
        // If we were woken up we'll just proceed - we typically only need to wait less than 10ms
    }
}
```

and then calling it via:
```
ThreadUtils.queue {
    (context as ApplicationContext).wakeUpDeviceAndDismissKeyguardIfRequired()
}
```

doesn't seem to actually wait at all. So with the above in place, if the app isn't visible we don't get any notifications and calls don't work, because the app is in the background so we can't restart a background or a foreground service - and instead we get this:
```
13:31:21.904  E  Failed to setCallInProgressNotification as a foreground service for type: 1, trying to update instead (Ask Gemini)
                 android.app.ForegroundServiceStartNotAllowedException: Service.startForeground() not allowed due to mAllowStartForeground false: service network.loki.messenger/org.thoughtcrime.securesms.service.WebRtcCallService```

Reply:
Ah, I think I misunderstood - rather than running the method on `ThreadUtils.queue` I should have just ran the "emergency exit" timer on `ThreadUtils.queue` - that works a lot better :)

Reply:
Hmm you didn't misunderstand, you got what I meant int he first place. The changes you have made: "run the waiting separately in the queue" - it means you are no longer "waiting" for it though, if this works, does that mean you don't have to wait in the first place?

Reply:
> Hmm you didn't misunderstand, you got what I meant int he first place. The changes you have made: "run the waiting separately in the queue" - it means you are no longer "waiting" for it though, if this works, does that mean you don't have to wait in the first place?

For references I was talking about this:

```java
// Wait for the device to actually wake up
            ThreadUtils.queue(() -> {
                long MAX_WAIT_PERIOD_MS = 50L;
                final long start = SystemClock.uptimeMillis();
                while (!powerManager.isInteractive() && SystemClock.uptimeMillis() - start <= MAX_WAIT_PERIOD_MS) {
                    try {
                        Thread.sleep(10);
                    } catch (InterruptedException e) {
                        // If we were woken up we'll just proceed - we typically only need to wait less than 10ms
                    }
                }
            });
```

Reply:
Just tested without any wakeup busy-waiting at all and it seems to work okay - which is odd, because if we busy wait with the entire method running via `ThreadUtils.queue` we don't get any call notifications when the app has been force-closed and the device locked.

Until we do some major work on calls (which Kee has said he'd like to do sometime in the new year) then we're still going to be a bit hit-and-miss with the call actually connecting, but at least with this patch we get to see the incoming call notifications so we have the opportunity to attempt to make the connection.

Reply:
All busy-waiting removed - resolving.